[OEP 3904]: online diskpool expansion over the same underlying device (#4031)
openebs/openebs: designs/replicated-pv/mayastor/pool-expansion.md

---
oep-number: OEP 3904
title: OpenEBS Enhancement Proposal for Mayastor DiskPool Expansion
authors:
  - "@abhilashshetty04"
owners:
  - "@abhilashshetty04"
editor: TBD
creation-date: 2025-08-13
last-updated: 2025-08-13
status: provisional
---

# OpenEBS Enhancement Proposal for Mayastor DiskPool Expansion

## Table of Contents

1. [Overview](#overview)
2. [Motivation](#motivation)
3. [Goals](#goals)
4. [Non-Goals](#non-goals)
5. [Proposal](#proposal)
6. [User Stories](#user-stories)
7. [Implementation Details](#implementation-details)
8. [Testing](#testing)

---

## Overview

> This proposal introduces the DiskPool Expansion feature to Mayastor pools of type **Lvs**, allowing users to increase pool capacity on the fly after expanding the underlying device.

Reviewer: Let's call it disk uniformly.

```suggestion
This proposal introduces the DiskPool Expansion feature to Mayastor pools of type **Lvs**, allowing users to increase pool capacity on the fly after expanding the underlying disk.
```

Author: disk seems to indicate to hdd type media. But we have already used disk a lot in the product. Ill make it disk.

## Motivation

1. Mayastor allows pool overcommitment via thin provisioning. In such cases, it is expected that as the allocation nears it's capacity, the storage administrator will proactively expand it to avoid replicas reaching the ENOSPC (no space) state. Currently, Mayastor does not support pool expansion.

2. Allows cluster to grow in Storage Capacity without adding new Pool or Node. It's an important utility for both on-premise or cloud based deployments if Pools are backed by storage devices which allows expansion.

3. We have received several user requests to support Mayastor pool expansion. Example:
   [User Request](https://github.com/orgs/openebs/projects/78?pane=issue&itemId=92928609&issue=openebs%7Cmayastor%7C1631)

> ## Goals

Reviewer 2: ```suggestion
## Goals

```

> - Allow user to grow the DiskPool upon expanding the device backing it.

Reviewer: ```suggestion
- Allow user to grow the DiskPool upon expanding the backing disk.
```

- Pool should be expandable when there are active IO operations. Without needing maintenance window.

## Non-Goals

- This proposal does not implement backend device grow capability. User will have to do that first before initiating mayastor DiskPool grow.
- This proposal does not implement Diskpool shrink.
- This proposal does not implement Expansion by adding additional Disk to the Pool.

> - This proposal does not implement infinite Grow capability. User has to define the expected growth ratio at DiskPool creation time.

Reviewer: ```suggestion
- This proposal does not implement infinite Grow capability. User has to define the expected growth ratio or expected growth size limit at DiskPool creation time.
```

## Proposal

1. **New CLI command** in the OpenEBS plugin:

>   kubectl openebs mayastor grow pool <pool-id>

Reviewer 2: We tend to use the term expansion (even here on the OEP title and overview), but then on the apis we use grow. Might be worth to keep the same term:
expand-pool: true
max_expansion: 
etc

Author: @avishnu , @dsharma-dc , @Abhinandan-Purkait  @niladrih shall we change it..

`expand-pool: true`... fine

`max-expansion` replaces `max-growth`

What about `max-growable-size`. Shall we keep it as is or should it be `expandable-upto` or something else? I think we should keep it as it. Its clear enough.

Reviewer: I'd say user facing naming can be made to use term `expand/expansion`, e.g the plugin command. Changing in APIs now might seem odd because there will be one named `GrowPool` and other named `ExpandPool` ? In contrast of what we have now with `GrowPool` and `GrowPoolV2`.

Author: >> I'd say user facing naming can be made to use term expand/expansion, e.g the plugin command. Changing in APIs now might seem odd because there will be one named GrowPool and other named ExpandPool ? In contrast of what we have now with GrowPool and GrowPoolV2.

Should be fine i guess. In code it will be `grow` throughout. @tiagolobocastro

Reviewer 2: If user facing is to become expand, then also both REST and CRD should be expand.
If we prefer keeping io-engine api as grow then it should be fine, though I don't see any issues with renaming it to ExpandPool, and this way we don't need the awkward V2 :) But fine either way provided the upper API is kept consistent

Author: Rest, crd and plugin can be expand.

`max-expansion` replaces `max-growth`

Should we replace `max-growable-size` with `expandable-upto`?

2. **DSP operator-based trigger**:

  Adding annotation `"openebs.io/grow-pool": true` on DSP CR will trigger the grow api for the Pool.

3. **New field in Pool CR spec**:

>   `max_growth` defines the maximum expected growth.

Reviewer 3: nit: This should be maxGrowth, if we're following camelCase for other names..

Reviewer 2: yeah, on the CR should be camel case

  Users can specify it in two ways:
  - **Absolute size**:
    - For ex. 800GiB, 3.5TiB or 4563726394B. User will have to specify the unit. Otherwise we will fail the request as it creates ambiguity.
    - Our implementation uses binary storage units, so values are expected in MiB, GiB, or TiB rather than MB, GB, or TB.
  - **Factors**:

>     - For ex. "1x", "20x", "8x" or "5x". 5x will allow the Pool to be grown 5 times of it's initial capacity. If not specified, We default it to '2x'.

Reviewer 3: 1x is not going to do anything, correct?

Author: Nope, its just going to be default reservation.

4. A New Pool/DSP status field

>   - `max_growable_size`: This is the absolute maximum growable size. Even growth of additional 1 cluster would fail. Please note that this would not impact Active Application IO. It's just that the extended capacity is unusable.

Reviewer: ```suggestion
  - `max_growable_size`: This is the absolute maximum growable size. Even growth of additional 1 cluster would fail. Please note that this would not impact any existing blobs or the volumes using those blobs for application IO. It's just that the extended backing disk capacity is unusable.
```

### Workflow

1. **Diskpool CRD migration**:
   - After the mayastor upgrade to release with this feature support. The existing CR will get a new field called `max_growable_size` in their status. Newer pool creation can have `max_growth` field in the spec to allow future growth.

2. **Diskpool Creation**:
    - The user/admin creates a diskpool yaml spec that contains the `max_growth` field indicating the expected growth in factor or absolute size.
    - When the spec is applied, the diskpool operator picks up this request and dispatches the create pool request to the Mayastor agents that complete the diskpool provisioning.
    - The spec for creating a pool with `max_growth` will look like below:

```
apiVersion: "openebs.io/v1beta3"
kind: DiskPool
metadata:
  name: <pool-name>
  namespace: <namespace>
spec:
  node: <node-name>
  disks: ["/dev/disk/by-id/<id>"]

>   max_growth: "20x" or "6TiB"

Reviewer: For the sake of clarity, I suggest providing two separate yamls here, one showing max_growth by factor, another showing max_growth by absolute value.

```
   - The User/Admin can check `max_growable_size` of the newly provisioned pool. It should be honoring the `max_growth`.

3. **Grow the DiskPool**:
   - The User/Admin must expand the backing device (≤ `max_growable_size`).
   - Then User/Admin can do either of the following to trigger mayastor pool grow workflow:
     - **Annotate the DSP CR**
       - add annotation `"openebs.io/grow-pool": true` on the CR to trigger reconciliation loop. The Operator removes the annotation when the grow succeeds or it sees error
       which is deemed irrecoverable.
     - **Use openebs plugin**
       - Run kubectl openebs mayastor grow pool <pool-id> which will call grow pool api.

## User Stories

1. **Story 1**:

>   As a System Administrator, I notice few Pool's Allocation is nearing Capacity. It has higher committed size then capacity. Hence, I want to grow the pool to avoid thin provisioned replicas encountering ENOSPC.

Reviewer 3: Is it a good idea to be able to check if pools are nearing their capacity using the plugin?

Author: Yes, users can do that. They can compare Allocated and Capacity They can also use monitoring stack.

## Implementation Details

### Design

- **Pool Creation workflow**:
   - The user/admin creates a diskpool yaml spec that contains the `max_growth` field indicating the expected growth.
   - As `max_growth` increases, depending on host performance we have noticed create pool takes more time due to higher number of metadata pages
   - Using larger `cluster_size` seems to reduce the time, As it reduces a number of metadata pages getting reserved, page can track more number of cluster too so `max_growth` is honoured.

>    - DSP Operator calls a grow_pool api, Request is passed down to the Io-Engine container present on the node which creates the pool with required configuration.

Reviewer: ```suggestion
   - DSP Operator calls a grow_pool api, request is passed down to the io-Engine container present on the node which creates the pool with required configuration.
```

   - Pool is created with enough metadata to grow the pool at least `max_growth` times it's initial capacity.
   - Once a pool is created, it's `max_growth` cannot be modified, and the `max_growable_size` remains fixed for the lifetime of the pool.

- **Pool Grow workflow**:
  - When a User/Admin triggers grow pool request. In io-engine, the underlying aio bdev is scanned. If we don't see extended `disk_capacity` we will return `FailedPrecondition` error. Will remove annotation if triggered from DSP operator.

>   - If metadata is exhausted because device is grown beyond `max_growable_size`, We will return ResourceExhausted. In this case also we will remove annotation if triggered from DSP Operator.

Reviewer 3: Why remove the annotation? Why not report failure in the CR status?

The storage admin may have issued the cloud API request to expand the device, and the device expansion is taking longer than usual. The storage admin then adds the annotation on DSP. But, due to ResourceExhausted, the annotation is removed. If the concern here is that some of the operations are costly, maybe a configurable/non-configurable retry frequency could be useful here?

A lot of cloud software uses the model of eventual consistency withing some predefined duration ceiling. A hard halt seems out of place? Maybe the user should be burdened with the task of removing the annotation?

Author: Growing the backend disk is a condition that user/admin must meet before starting the expansion.
 It could also be a case where user has not triggered request to expand the disk and uses our expand utility.
 
The other error is metadata exhaustion. That is irrecoverable. So want to remove annotations there also. Don't want to subscribe to declarative approach where its not needed. These are about meeting prerequisites before calling expansion. Declarative approach seems better when its runtime scenario handling.

We can add status to dsp cr. We can also emit events from io-engine itself.

Reviewer 3: Ok

- This proposal does not implement Diskpool shrink.
- This proposal does not implement Expansion by adding additional Disk to the Pool.
- This proposal does not implement infinite Grow capability. User has to define the expected growth ratio at DiskPool creation time.

## Proposal

1. **New CLI command** in the OpenEBS plugin:

  kubectl openebs mayastor grow pool <pool-id>

2. **DSP operator-based trigger**:

  Adding annotation `"openebs.io/grow-pool": true` on DSP CR will trigger the grow api for the Pool.

3. **New field in Pool CR spec**:

  `maxGrowth` defines the maximum expected growth.

  Users can specify it in two ways:
  - **Absolute size**:
    - For ex. 800GiB, 3.5TiB or 4563726394B. User will have to specify the unit. Otherwise we will fail the request as it creates ambiguity.
    - Our implementation uses binary storage units, so values are expected in MiB, GiB, or TiB rather than MB, GB, or TB.
  - **Factors**:
    - For ex. "1x", "20x", "8x" or "5x". 5x will allow the Pool to be grown 5 times of it's initial capacity. If not specified, We default it to '2x'.

4. A New Pool/DSP status field

>   - `max_growable_size`: This is the absolute maximum growable size. Even growth of additional 1 cluster would fail. Please note that this would not impact Active Application IO. It's just that the extended capacity is unusable.

Reviewer: All mentions of `max_growable_size` now need to say `max_expandable_size` now, right?

  - `disk_capacity` will show the size of the underlying disk.

  NOTE: `capacity` is the actual usable space of the Pool after reserving metadata from `disk_capacity`. So while growing user/admin should consider current `disk_capacity` and `max_growable_size`

### Key Concepts

**LVS Metadata restriction**:
  When a mayastor pool is created, The metadata region is initialized which is immutable. Future expansion of the pool depends on the metadata reserved on create time. `maxGrowth` in the CR spec is used to reserve required metadata. All DiskPools created prior to this will have a limited grow capability as they would be not having enough metadata to accommodate bigger disk size. The User/Admin might need to manually move replica off of the Pool and recreate the Pool by using `maxGrowth` if they want to grow older Pools.

### Workflow

1. **Diskpool CRD migration**:
   - After the mayastor upgrade to release with this feature support. The existing CR will get a new field called `max_growable_size` in their status. Newer pool creation can have `maxGrowth` field in the spec to allow future growth.

2. **Diskpool Creation**:

>     - The user/admin creates a diskpool yaml spec that contains the `maxGrowth` field indicating the expected growth in factor or absolute size.

Reviewer 2: ```suggestion
    - The user/admin creates a diskpool yaml spec that contains the `maxExpansion` field indicating the expected growth in factor or absolute size.
```
